src/java/htsjdk/samtools/util/AsyncBlockCompressedInputStream.java

```diff
@@ -0,0 +1,191 @@
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package htsjdk.samtools.util;
+
+
+import htsjdk.samtools.Defaults;
+import htsjdk.samtools.seekablestream.SeekableStream;
+
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.URL;
+import java.util.concurrent.ArrayBlockingQueue;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.Executor;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Semaphore;
+import java.util.concurrent.ThreadFactory;
+
+/**
+ * Asynchronous read-ahead implementation of {@link htsjdk.samtools.util.BlockCompressedInputStream}.   
+ * 
+ * Note that this implementation is not synchronized. If multiple threads access an instance concurrently, it must be synchronized externally. 
+ */
+public class AsyncBlockCompressedInputStream extends BlockCompressedInputStream {
+    private static final int READ_AHEAD_BUFFERS = (int)Math.ceil(Defaults.NON_ZERO_BUFFER_SIZE / BlockCompressedStreamConstants.MAX_COMPRESSED_BLOCK_SIZE);
+    private static final Executor threadpool = Executors.newFixedThreadPool(Runtime.getRuntime().availableProcessors(),new ThreadFactory() {
+            public Thread newThread(Runnable r) {
+                Thread t = Executors.defaultThreadFactory().newThread(r);
+                t.setDaemon(true);
+                return t;
+            }
+        });
+    private final BlockingQueue<DecompressedBlock> mResult = new ArrayBlockingQueue<>(READ_AHEAD_BUFFERS);
+    private final BlockingQueue<byte[]> freeBuffers = new ArrayBlockingQueue<>(READ_AHEAD_BUFFERS);
+    private final Semaphore running = new Semaphore(1);
+    private volatile boolean mAbort = false;
+    private AsyncBlockCompressedInputStreamRunnable runnable = new AsyncBlockCompressedInputStreamRunnable();
+
+    public AsyncBlockCompressedInputStream(final InputStream stream) {
+        super(stream, true);
+    }
+
+    public AsyncBlockCompressedInputStream(final File file)
+        throws IOException {
+        super(file);
+    }
+
+    public AsyncBlockCompressedInputStream(final URL url) {
+        super(url);
+    }
+
+    public AsyncBlockCompressedInputStream(final SeekableStream strm) {
+        super(strm);
+    }
+    
+    protected DecompressedBlock nextBlock(byte[] bufferAvailableForReuse) {
+        if (bufferAvailableForReuse != null) {
+            freeBuffers.offer(bufferAvailableForReuse);
+        }
+        return nextBlockSync();
+    }
+    
+    @Override
+    protected void prepareForSeek() {
+        flushReadAhead();
+        super.prepareForSeek();
+    }
+
+    @Override
+    public void close() throws IOException {
+        // Suppress interrupts while we close.
+        final boolean isInterrupted = Thread.interrupted();
+        mAbort = true;
+        flushReadAhead();
+        super.close();
```

Comment:
what if this close blows up (as it can by the IOException in the signature)? the interrupt status will not be restored, right?